extensions/annotations/src/main/java/io/opentelemetry/extension/annotations/Timed.java

```diff
@@ -0,0 +1,101 @@
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This annotation creates a {@link io.opentelemetry.api.metrics.LongHistogram Histogram} instrument
+ * observing the duration of invocations of the annotated method or constructor.
+ *
+ * <p>By default, the Histogram instrument will have the following attributes:
+ *
+ * <ul>
+ *   <li><b>class:</b> The fully qualified name of the class whose method is invoked.
+ *   <li><b>method:</b> The name of the annotated method, or "new" of the annotation is on a
+ *       constructor.
+ *   <li><b>exception:</b> The {@link Class#getSimpleName() simple name} of the Exception if an
+ *       Exception is thrown, or "None" if the method did not throw an Exception.
+ * </ul>
+ *
+ * <p>Application developers can use this annotation to signal OpenTelemetry auto-instrumentation
+ * that the Histogram instrument should be created.
+ *
+ * <p>If you are a library developer, then probably you should NOT use this annotation, because it
+ * is non-functional without the OpenTelemetry auto-instrumentation agent, or some other annotation
+ * processor.
+ */
+@Target({ElementType.METHOD, ElementType.CONSTRUCTOR})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Timed {
+
+  /**
+   * Name of the Histogram instrument.
+   *
+   * <p>The name should follow the instrument naming rule: <a
+   * href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-naming-rule">https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-naming-rule</a>
+   *
+   * <p>The default name is {@code method.invocations.duration}.
+   */
+  String value() default "method.invocations.duration";
+
+  /**
+   * Description for the instrument.
+   *
+   * <p>Description strings should follow the instrument description rules: <a
+   * href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-description">https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#instrument-description</a>
+   */
+  String description() default "";
+
+  /**
+   * The unit for the instrument.
+   *
+   * <p>Default is seconds.
+   */
+  TimeUnit unit() default TimeUnit.SECONDS;
+
+  /**
+   * List of key-value pairs to supply additional attributes.
+   *
+   * <p>Example:
+   *
+   * <pre>
+   * {@literal @}Timed(
+   *     additionalAttributes = {
+   *       "key1", "value1",
+   *       "key2", "value2",
+   * })
+   * </pre>
+   */
+  String[] attributes() default {};
+
+  /**
+   * Attribute name for the return value.
+   *
+   * <p>The name of the attribute for the return value of the method call. {@link Object#toString()}
+   * will be called on the return value to convert it to a String.
+   *
+   * <p>By default, the instrument will not have an attribute with the return value.
+   */
+  String returnValueAttribute() default "";
```

Comment:
A couple days ago somebody asked for the same thing for spans: https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/5527
Instead of an attribute, perhaps this could be a separate annotation? That's used by both signals (traces and metrics)

Reply:
There should be an option to use an attribute for a span but not for a metric. For metrics, each set of attributes creates a new time series, and users must be careful to use only attributes that have a limited number of possible different values. For spans that is not an issue. So it might make sense to use arbitrary parameters, return values, and exceptions as Span attributes but not use them as metric attributes.

Reply:
I guess if we want to have an option to use the return value as a span attribute but not as a metric attribute we don't need an extra annotation, because then we cannot reuse it for spans and metrics anyway. In that case we could just leave the `returnValueAttribute` parameter.

Reply:
I think @mateuszrzeszutek was suggesting something like this:

```
  @Timed
  @ReturnAttribute("process.result")
  public int process() {
    ...
  }
```

Reply:
Yeah, exactly that -- and this annotation could be used to fill in attributes both in spans and metrics.

Reply:
I'm not quite sure whether we should have common attributes for both spans and metrics, or whether we should separate them. The difference is that with metrics each new attribute value creates a new time series. Therefore you only want to choose attributes that have a limited number of possible values. With Spans that's not the case.

An example would be a method returning a unique identifier: It would be valid to use the return value as a Span attribute, but it would be a bad idea to use it as a metric attribute.

For method parameters the current proposal is to have two separate annotations: `@MetricAttribute` and `@SpanAttribute`. If you want a parameter to be used in both, metric and span, you would need to add both annotations.

If we want to have the same separation for return values, we would end up with something like `@MetricReturnAttribute` and `@SpanReturnAttribute`. That's fine. I find it shorter to have this as a parameter to the `@Counted`, `@Timed`, and maybe `@WithSpan`, but I have no strong opinion about this.

However, having a common `@ReturnAttribute` for both metrics and spans would make it impossible to use a high cardinality return value as a Span attribute without using it as a metric attribute at the same time.

What do you think?

Reply:
Oh, I understand now, yeah that makes sense that you would want control over which attributes are tagged on the span vs tagged on the metric